frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseDiscoverAction.java

```diff
@@ -0,0 +1,78 @@
+package com.virtuslab.gitmachete.frontend.actions.base;
+
+import static com.intellij.openapi.application.ModalityState.NON_MODAL;
+import static com.virtuslab.gitmachete.frontend.resourcebundles.GitMacheteBundle.getString;
+
+import java.nio.file.Path;
+
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.Task;
+import com.intellij.openapi.project.DumbAwareAction;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vcs.VcsNotifier;
+import com.intellij.ui.GuiUtils;
+import io.vavr.control.Try;
+import lombok.CustomLog;
+
+import com.virtuslab.binding.RuntimeBinding;
+import com.virtuslab.branchlayout.api.readwrite.IBranchLayoutWriter;
+import com.virtuslab.gitmachete.backend.api.IGitMacheteRepositoryCache;
+import com.virtuslab.gitmachete.backend.api.IGitMacheteRepositorySnapshot;
+import com.virtuslab.gitmachete.frontend.actions.dialogs.GraphTableDialog;
+import com.virtuslab.gitmachete.frontend.actions.expectedkeys.IExpectsKeyProject;
+import com.virtuslab.gitmachete.frontend.ui.api.table.BaseGraphTable;
+import com.virtuslab.gitmachete.frontend.ui.providerservice.SelectedGitRepositoryProvider;
+import com.virtuslab.gitmachete.frontend.vfsutils.GitVfsUtils;
+import com.virtuslab.logger.IEnhancedLambdaLogger;
+
+@CustomLog
+public class BaseDiscoverAction extends DumbAwareAction implements IExpectsKeyProject {
+  @Override
+  public void actionPerformed(AnActionEvent anActionEvent) {
+    var project = getProject(anActionEvent);
+    var selectedRepoProvider = project.getService(SelectedGitRepositoryProvider.class).getGitRepositorySelectionProvider();
+    var selectedRepositoryOption = selectedRepoProvider.getSelectedGitRepository();
+    assert selectedRepositoryOption.isDefined() : "Selected repository is undefined";
+    var selectedRepository = selectedRepositoryOption.get();
+    var mainDirPath = GitVfsUtils.getMainDirectoryPath(selectedRepository).toAbsolutePath();
+    var gitDirPath = GitVfsUtils.getGitDirectoryPath(selectedRepository).toAbsolutePath();
+    Try.of(() -> RuntimeBinding.instantiateSoleImplementingClass(IGitMacheteRepositoryCache.class)
```

Comment:
Maybe make the cache instance a field (potentially a `Lazy` field) of `BaseDiscoverAction`

Reply:
Although this is of little significance anyway since this action isn't going to be invoked more than once on a given project anyway... up to you

Reply:
Yeah, this action likely to be not executed too frequently, so I leave as is